### PR TITLE
Jdbc Operator How To

### PR DESCRIPTION
**Pull request by francescomucio**
Added an How To page for the Jdbc operator (and an example dag to get code snippet for the page)
 
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: docs/howto/operator/jdbc.rst [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+each command (default is ``False``)
+
+Prerequisite Tasks
+------------------
+To use this operator you need:
+
+  * Install the python module jaydebeapi:
+    .. code-block:: bash
+
+      pip install jaydebeapi
```
Comment (mik-laj) [inline docs/howto/operator/jdbc.rst]: I think, we should use extra packages to manage dependencies for Airflow.  WDYT?
See: https://airflow.readthedocs.io/en/latest/installation.html#extra-packages

File: docs/howto/operator/jdbc.rst [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    :end-before: [END howto_operator_jdbc]
+
+The parameter ``sql`` can receive a string or a list of strings.
+Each string can be an SQL statement or a reference to a template file.
+Template reference are recognized by ending in '.sql'.
+
+The parameter ``autocommit`` if set to ``True`` will execute a commit after
+each command (default is ``False``)
+
+Prerequisite Tasks
```
Comment (mik-laj) [inline docs/howto/operator/jdbc.rst]: This should be the second section in the document and the first should describe the JDBC.

File: docs/howto/operator/jdbc.rst [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+------------------
+To use this operator you need:
+
+  * Install the python module jaydebeapi:
+    .. code-block:: bash
+
+      pip install jaydebeapi
+
+  * Install a JVM.
```
Comment (mik-laj) [inline docs/howto/operator/jdbc.rst]: Can you add a link to external docs about it?

File: docs/howto/operator/jdbc.rst [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  * Have the JDBC driver for your database installed.
```
Comment (mik-laj) [inline docs/howto/operator/jdbc.rst]: Can you add a link to external docs about it?  Ideally, you should include a simple command here to check if this condition is met.

File: docs/howto/operator/jdbc.rst (hunk 1 above)
Comment (francescomucio) [reply docs/howto/operator/jdbc.rst]: In the software section like for ODBC?
Should I add this as part of this PR?

File: docs/howto/operator/jdbc.rst (hunk 3 above)
Comment (francescomucio) [reply docs/howto/operator/jdbc.rst]: I always found installing Java very platform specific, what do you think of https://adoptopenjdk.net/installation.html?
I guess it is missing the step about setting the JAVA_HOME folder like you do in [your Dockerfile](https://gist.github.com/mik-laj/c090f51c970e4f6898e6fd986ea1e12e#file-java-hadoop-dockerfile-L18).

File: docs/howto/operator/jdbc.rst (hunk 2 above)
Comment (francescomucio) [reply docs/howto/operator/jdbc.rst]: moved

File: docs/howto/operator/jdbc.rst (hunk 4 above)
Comment (francescomucio) [reply docs/howto/operator/jdbc.rst]: I have included a simple check, not sure if it is enough

File: airflow/example_dags/example_jdbc_operator.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
+#
```
Comment (kaxil) [inline airflow/example_dags/example_jdbc_operator.py]: This file should be in `airflow.providers.jdbc.example_dags`

File: airflow/example_dags/example_jdbc_operator.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+) as dag:
+
+    run_this_last = DummyOperator(
+        task_id='run_this_last',
+        dag=dag,
+    )
+
+    # [START howto_operator_jdbc_template]
+    delete_data = JdbcOperator(
+        task_id='insert_data',
```
Comment (mik-laj) [inline airflow/example_dags/example_jdbc_operator.py]:
```suggestion
        task_id='delete_data',
```

File: docs/howto/operator/jdbc.rst [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+Prerequisite Tasks
+^^^^^^^^^^^^^^^^^^
+
+To use this operator you need:
+
+  * Install the python module jaydebeapi:
+    .. code-block:: bash
+
+      pip install jaydebeapi
```
Comment (mik-laj) [inline docs/howto/operator/jdbc.rst:39]:
```suggestion
      pip install 'apache-airflow[jdbc]'
```

File: docs/operators-and-hooks-ref.rst [hunk 8]
```diff
@@ -75,6 +75,8 @@ Fundamentals
 
    * - :mod:`airflow.operators.latest_only`
      -
+   * - :mod:`airflow.providers.jdbc.operators.jdbc`
```
Comment (mik-laj) [inline docs/operators-and-hooks-ref.rst]: The fundamentals section only includes operators that are in the Airflow core. Please look at the protocol section: https://airflow.readthedocs.io/en/latest/operators-and-hooks-ref.html#protocol-operators-and-hooks

File: airflow/example_dags/example_jdbc_operator.py (hunk 5 above)
Comment (francescomucio) [reply airflow/example_dags/example_jdbc_operator.py]: moved

File: airflow/example_dags/example_jdbc_operator.py (hunk 6 above)
Comment (francescomucio) [reply airflow/example_dags/example_jdbc_operator.py]: thanks, fixed

File: docs/howto/operator/jdbc.rst (hunk 7 above)
Comment (francescomucio) [reply docs/howto/operator/jdbc.rst:39]: modified

File: docs/operators-and-hooks-ref.rst (hunk 8 above)
Comment (francescomucio) [reply docs/operators-and-hooks-ref.rst]: Thanks for pointing out this, I removed this and moved the link to the JDBC doc below

**Review by mik-laj — APPROVED** [review]

**Comment by kaxil** [comment]
Good work @francescomucio 🎉

**Comment by ashb** [comment]
This, I think, has broken master and all open PRs

```
>       self.assertEqual([], invalid_files)
E       AssertionError: Lists differ: [] != ['/opt/airflow/airflow/providers/jdbc/example_dags/example_jdbc_operator.py']
E       
E       Second list contains 1 additional elements.
E       First extra element 0:
E       '/opt/airflow/airflow/providers/jdbc/example_dags/example_jdbc_operator.py'
E       
E       - []
E       + ['/opt/airflow/airflow/providers/jdbc/example_dags/example_jdbc_operator.py']

tests/core/test_project_structure.py:230: AssertionError
```

**Comment by ashb** [comment]
@potiuk Selective tests (? or something else) caused CI to not detect failing tests:

https://github.com/apache/airflow/runs/1307780299

```
tests/core/test_project_structure.py sssssssssss                         [ 23%]
```

**Comment by francescomucio** [comment]
Sorry for the mess, not sure what I should keep an eye on next time

**Pull request merged**